File: Embarcadero/Berlin/CPP/ThingPoint IoT Sample Data Demo/ThingpointCpp/StreamingDeviceFrameU.cpp
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
#include <fmx.h>
#pragma hdrstop

#include "StreamingDeviceFrameU.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TStreamingDeviceFrame *StreamingDeviceFrame;
//---------------------------------------------------------------------------
__fastcall TStreamingDeviceFrame::TStreamingDeviceFrame(TComponent* Owner)
	: TFrame(Owner)
{
  FValue = new TJSONObject();
}
//---------------------------------------------------------------------------
void __fastcall TStreamingDeviceFrame::ActionStartStopExecute(TObject * Sender)
{
  if (Started)
    Stop();
  else
    Start();
}

void __fastcall TStreamingDeviceFrame::ActionStartStopUpdate(TObject * Sender)
{
  if (Started)
    ((TAction*)Sender)->Text = "Stop";
  else
    ((TAction*)Sender)->Text = "Start";
}

void __fastcall TStreamingDeviceFrame::ButtonClearClick(TObject * Sender)
{
  DoClear();
}

void __fastcall TStreamingDeviceFrame::ClearLog(void)
{
  Memo1->Lines->Clear();
}

void __fastcall TStreamingDeviceFrame::DoClear(void)
{
  if (FOnClear != NULL)
    FOnClear(this);
}

__fastcall TStreamingDeviceFrame::~TStreamingDeviceFrame()
{
  delete FValue;
}

void __fastcall TStreamingDeviceFrame::DoChanged(void)
{
  if (FOnChanged != NULL)
    FOnChanged(this);
}

void __fastcall TStreamingDeviceFrame::DoNextValue(void)
{
  if (this->FOnNextValue != NULL)
  {
    while (FValue->Count > 0)
      FValue->RemovePair(FValue->Pairs[0]->JsonString->Value());
    FOnNextValue(FValue);
  }
}

bool __fastcall TStreamingDeviceFrame::GetStarted(void)
{
  return Timer1->Enabled;
}

void __fastcall TStreamingDeviceFrame::LogValue(void)
{
  Memo1->Lines->Insert(0, String::Format("[%s] %s", ARRAYOFCONST((FormatDateTime("mm:ss:zzz", FTime), Value->ToString()))));
}

void __fastcall TStreamingDeviceFrame::NextValue(void)
{
  DoNextValue();
  FTime = Now();
}

void __fastcall TStreamingDeviceFrame::SetStarted(bool Value)
{
  Timer1->Enabled = Value;
}

void __fastcall TStreamingDeviceFrame::Start(void)
{
  Timer1->Enabled = true;
}

void __fastcall TStreamingDeviceFrame::Stop(void)
{
  Timer1->Enabled = false;
}

void __fastcall TStreamingDeviceFrame::Timer1Timer(TObject * Sender)
{
  NextValue();
  DoChanged();
}

File: Embarcadero/Berlin/Object Pascal/ThingPoint IoT Sample Data Demo/Thingpoint/StreamingDeviceFrameU.pas
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
unit StreamingDeviceFrameU;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants, 
  FMX.Types, FMX.Graphics, FMX.Controls, FMX.Forms, FMX.Dialogs, FMX.StdCtrls,
  System.Actions, FMX.ActnList, FMX.Controls.Presentation, FMX.ScrollBox,
  FMX.Memo, FMX.Layouts, System.JSON;

type
  TStreamingDeviceFrame = class(TFrame)
    LabelDevice: TLabel;
    ButtonStartStop: TButton;
    ActionList1: TActionList;
    ActionStartStop: TAction;
    Timer1: TTimer;
    Memo1: TMemo;
    Layout1: TLayout;
    Layout2: TLayout;
    Layout3: TLayout;
    Layout4: TLayout;
    ButtonClear: TButton;
    procedure ActionStartStopExecute(Sender: TObject);
    procedure ActionStartStopUpdate(Sender: TObject);
    procedure Timer1Timer(Sender: TObject);
    procedure ButtonClearClick(Sender: TObject);
  public type
    TNextValueCallback = reference to procedure(const Value: TJSONObject);
  private
    FValue: TJSONObject;
    FTime: TDateTime;
    FOnChanged: TNotifyEvent;
    FNextValueCallback: TNextValueCallback;
    FOnClear: TNotifyEvent;
    procedure SetStarted(const Value: Boolean);
    function GetStarted: Boolean;
    procedure DoClear;
    { Private declarations }
  protected
    procedure DoChanged; virtual;
    procedure DoNextValue; virtual;
  public
    { Public declarations }
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
    procedure ClearLog;
    procedure Start;
    procedure Stop;
    procedure LogValue;
    procedure NextValue;
    property Value: TJSONObject read FValue;
    property Time: TDateTime read FTime;
    property Started: Boolean read GetStarted write SetStarted;
    property OnChanged: TNotifyEvent read FOnChanged write FOnChanged;
    property OnClear: TNotifyEvent read FOnClear write FOnClear;
    property OnNextValue: TNextValueCallback write FNextValueCallback;
  end;

implementation

{$R *.fmx}

procedure TStreamingDeviceFrame.ActionStartStopExecute(Sender: TObject);
begin
  if Started then
    Stop
  else
    Start;
end;

procedure TStreamingDeviceFrame.ActionStartStopUpdate(Sender: TObject);
begin
  if Started then
    TAction(Sender).Text := 'Stop'
  else
    TAction(Sender).Text := 'Start'
end;

procedure TStreamingDeviceFrame.ButtonClearClick(Sender: TObject);
begin
  DoClear;
end;

procedure TStreamingDeviceFrame.ClearLog;
begin
  Memo1.Lines.Clear;
end;

procedure TStreamingDeviceFrame.DoClear;
begin
  if Assigned(FOnClear) then
    FOnClear(Self);
end;

constructor TStreamingDeviceFrame.Create(AOwner: TComponent);
begin
  inherited;
  FValue := TJSONObject.Create;
end;

destructor TStreamingDeviceFrame.Destroy;
begin
  FValue.Free;
  inherited;
end;

procedure TStreamingDeviceFrame.DoChanged;
begin
  if Assigned(FOnChanged) then
    FOnChanged(Self);
end;

procedure TStreamingDeviceFrame.DoNextValue;
begin
  if Assigned(Self.FNextValueCallback) then
  begin
    while FValue.Count > 0 do
      FValue.RemovePair(FValue.Pairs[0].JsonString.Value);
    FNextValueCallback(FValue);
  end;
end;

function TStreamingDeviceFrame.GetStarted: Boolean;
begin
  Result := Timer1.Enabled;
end;

procedure TStreamingDeviceFrame.LogValue;
begin
  Memo1.Lines.Insert(0, Format('[%s] %s', [FormatDateTime('mm:ss:zzz', FTime), Value.ToString]));
end;

procedure TStreamingDeviceFrame.NextValue;
begin
  DoNextValue;
  FTime := Now;
end;

procedure TStreamingDeviceFrame.SetStarted(const Value: Boolean);
begin
  Timer1.Enabled := Value;
end;

procedure TStreamingDeviceFrame.Start;
begin
  Timer1.Enabled := True;
end;

procedure TStreamingDeviceFrame.Stop;
begin
  Timer1.Enabled := False;
end;

procedure TStreamingDeviceFrame.Timer1Timer(Sender: TObject);
begin
  NextValue;
  DoChanged;
end;

end.
